DirectProgramming/C++SYCL_FPGA/Tutorials/DesignPatterns/state_machine/src/state_machine.hpp

```diff
@@ -0,0 +1,133 @@
+                                         StreamingPipeProps>;
+using StreamIn_OptimizedStateMachine =
+    sycl::ext::intel::experimental::pipe<IDStreamDataIn_OptimizedStateMachine, MyStreamingBeat, 0,
+                                         StreamingPipeProps>; 
+using StreamOut_NaiveStateMachine =
+    sycl::ext::intel::experimental::pipe<IDStreamDataOut_NaiveStateMachine, MyStreamingBeat, 0,
+                                         StreamingPipeProps>;
+using StreamOut_OptimizedStateMachine =
+    sycl::ext::intel::experimental::pipe<IDStreamDataOut_OptimizedStateMachine, MyStreamingBeat, 0,
+                                         StreamingPipeProps>;
+
+// function for task_sequence
+float UpdateCoeff(float *coeff_ptr, int *execution_count) {
+  float coeff_new = coeff_ptr[0];
+  // read-modify-write of external memory will result in a poor II.
+  execution_count[0] = execution_count[0] + 1;
+  return coeff_new;
+}
+
+// function for task_sequence
+float Compute(float coeff, float data) {
+  return coeff * data;
+}
+
+struct NaiveStateMachine {
+  float *coeff_ptr;
+  int *execution_count;
+  bool init_coeff_before_starting = false;
+  int iterations;
+
+  void operator()() const {
+    State my_state = 
+        (init_coeff_before_starting) ? State::LD_COEFF : State::PROCESS;
+    float coeff = 1.0f;
+
+    for (int i = 0; i < iterations; i++) {
+      switch (my_state) {
+        case State::PROCESS: {
+          MyStreamingBeat beat = StreamIn_NaiveStateMachine::read();
+          // multiplication of input data with a coefficient can occur with II=1
+          beat.data = beat.data * coeff;
+          StreamOut_NaiveStateMachine::write(beat);
+        } break;
+        
+        case State::LD_COEFF: {
+          float coeff_new = coeff_ptr[0];
+          execution_count[0] = execution_count[0] + 1;
+          coeff = coeff_new;
+          my_state = State::PROCESS;
+        } break;
+        
+        default:
+          break;
+      }
+    }
+  }
+};
+
+struct OptimizedStateMachine {
+  float *coeff_ptr;
+  int *execution_count;
+  bool init_coeff_before_starting = false;
+  int iterations;
+
+  void operator()() const {
+    State my_state =
+        (init_coeff_before_starting) ? State::LD_COEFF : State::PROCESS;
+    float coeff = 1.0f;
+
+    sycl::ext::intel::experimental::task_sequence<UpdateCoeff> update_coeff_task;
+    sycl::ext::intel::experimental::task_sequence<Compute> compute_task;
+
+    [[intel::initiation_interval(1)]] // No-format: Attribute
+    for (int i = 0; i < iterations; i++) {
+      switch (my_state) {
+        case State::PROCESS: {
+          MyStreamingBeat beat = StreamIn_OptimizedStateMachine::read();
+          // use task_sequence to hide long II from compiler
+          compute_task.async(coeff, beat.data);
+          beat.data = compute_task.get();
+          StreamOut_OptimizedStateMachine::write(beat);
```

Comment:
```suggestion
          // This state should achieve II=1
          compute_task.async(coeff);
```
If we move **all** the processing from the state machine into the task function, then we can remove the `get()` call. Does this allow II=1 during simulation?

Reply:
Paul's suggestion should theoretically improve the II since you will not have the async-to-get dependence that will create the problem with capacity balancing.

Reply:
Understood. Thanks for pointing that out. However, Johnny suggested the same and we actually worked on several workarounds. Some of the ways went too hacky and introduced too much extra code just for this simple state machine. The II issue persisted regardless, which is very sad.